File: RB110/ruby_interview_problems/new_rb119_interview_study_problems/problem3.rb
=begin
Create a method that takes a string argument and returns a copy of 
the string with every second character in every third word converted
to uppercase. Other characters should remain the same.

Examples:
original = 'Lorem Ipsum is simply dummy text of the printing world'
expected = 'Lorem Ipsum iS simply dummy tExT of the pRiNtInG world'
p to_weird_case(original) == expected

original = 'It is a long established fact that a reader will be distracted'
expected = 'It is a long established fAcT that a rEaDeR will be dIsTrAcTeD'
p to_weird_case(original) == expected

p to_weird_case('aaA bB c') == 'aaA bB c'

original = "Mary Poppins' favorite word is supercalifragilisticexpialidocious"
expected = "Mary Poppins' fAvOrItE word is sUpErCaLiFrAgIlIsTiCeXpIaLiDoCiOuS"
p to_weird_case(original) == expected

Problem:
- Return the given string with every second character in every 3rd word uppercased
- All other chars should remain the same
- Is a copy of the string a new string or the same given string?

Input: string
Output: modified string

Example:
original = "Mary Poppins' favorite word is supercalifragilisticexpialidocious"
- Every third word -> favorite, supercalifragi...
  - every 2nd Char -> fAvOrItE, sUpErCaLiFrAgI...
  -> "Mary Poppins' fAvOrItE word is sUpErCaLiFrAgIlIsTiCeXpIaLiDoCiOuS"

Data Structures:
- Array would allow to easily modify each word and convert from string to array
  and array to string

Algorithm:
- SPLIT the string into an array of words
- ITERATE over the words
  - for every third word, iterate over that word
    - every second character should be upcased
- JOIN the array back into a string, where words are separated by spaces

=end

def to_weird_case(string)
  string.split.map.with_index do |word, idx|
    if (idx + 1) % 3 == 0
      word.chars.map.with_index do |char, idx| 
        (idx + 1) % 2 == 0 ? char.upcase : char
      end.join
    else
      word
    end
  end.join(' ')
end

original = 'Lorem Ipsum is simply dummy text of the printing world'
expected = 'Lorem Ipsum iS simply dummy tExT of the pRiNtInG world'
p to_weird_case(original) == expected

original = 'It is a long established fact that a reader will be distracted'
expected = 'It is a long established fAcT that a rEaDeR will be dIsTrAcTeD'
p to_weird_case(original) == expected

p to_weird_case('aaA bB c') == 'aaA bB c'

original = "Mary Poppins' favorite word is supercalifragilisticexpialidocious"
expected = "Mary Poppins' fAvOrItE word is sUpErCaLiFrAgIlIsTiCeXpIaLiDoCiOuS"
p to_weird_case(original) == expected

File: LS215/JS_Small_Problems/LS215_Exam_Practice_Problems/problem3.js
// A collection of spelling blocks has two letters per block, as shown in this list:


// B:O   X:K   D:Q   C:P   N:A
// G:T   R:E   F:S   J:W   H:U
// V:I   L:Y   Z:M
// This limits the words you can spell with the blocks to only those words that do not use both letters from any given block. You can also only use each block once.

// Write a function that takes a word string as an argument and returns true if the word can be spelled using the set of blocks, false otherwise. You can consider the letters to be case-insensitive when you apply the rules.

/*
----- PROBLEM -----
INPUT: string
OUTPUT: boolean true or false

Rules:
  - Given a string, return true based on if the word can be spelled using the blocks where
    1. only 1 letter from the block pair can be used
    2. case does not apply
    3. eack pair of blocks can only be used once
    4. this implicitly means that the longest string we can have is 13 chars long
      since there are 13 blocks, one letter from each block

Assumptions:
  - Will the given string contain only alpha chars?
    - Return null if the string has nonalpha chars
  - What happens if passed an empty string?
    - Return null
  - What happens if string that is longer than 13chars is passed?
    - return false
  - What happens if a non string arg is passed?
    - return null

----- EXAMPLES -----
isBlockWord('BATCH');      // true
  - 5 char word `BATCH` -> true
    - B is paired with O
    - A is paired with N
    - T is paired with G
    - C is paired with P
    - H is paired with U
  - Since each letter is used once and each block is only used once, true is returned

isBlockWord('BUTCH');      // false
  - 5 char word `BUTCH` -> false, thus we have a block repeated more than 1x or or 2 letters frmo same block used
    - B is paired with O
    - U is paired with H <----
    - T is paired with G
    - C is paired with P
    - H is paired with U <----

  - since H and U are used which are from the same block, we return false


----- DATA STRUCTURES -----
- Object would allow easy way tos tore block pairs
- Meanwhile, array would allow easy method to check which block-pairs we've used


----- ALGORITHM -----
- Return null or false if invalid argument is passed
  - Check if arg is > 13 chars, contains non alpha chars, or is not a string arg

- CREATE constant obj w/ all 13 pairs of letters
- Add pair of letters into array for each letter in the given argument
  - Declare empty array `letters`
  - Iterate over string
  - For each character in string
    - Check the array to see if it includes the key (letter1) or value (letter2)
      - If so
        - return false
    - If not
      - Add the letters to the array and mvoe onto the next character in the str argument
- Return true
*/

const PAIRS = {
  'B': 'O', 'X': 'K', 'D': 'Q', 'C': 'P', 'N': 'A',
  'G': 'T', 'R': 'E', 'F': 'S', 'J': 'W', 'H': 'U',
  'V': 'I', 'L': 'Y', 'Z': 'M',
};

function isBlockWord(word) {
  if (/[^a-z]/ig.test(word) || typeof word !== 'string' || word === '') return null;
  if (word.length > 13) return false;
  word = word.toUpperCase();
  let letters = [];

  for (let idx = 0; idx < word.length; idx += 1) {
    for (let key in PAIRS) {
      if (key === word[idx] || PAIRS[key] === word[idx]) {
        if (letters.includes(key)) return false;
        letters.push(key, PAIRS[key]);
      }
    }
  }

  return true;
}


console.log(isBlockWord('BATCH'));          // true
console.log(isBlockWord('jest'));           // true
console.log(isBlockWord('BGVXRLDFZCJNH'));  // true
console.log(isBlockWord('BUTCH'));          // false
console.log(isBlockWord('BGVXRLDFZCJNHA')); // false
console.log(isBlockWord('B@TCH'));          // null
console.log(isBlockWord(1));                // null
console.log(isBlockWord(''));               // null